openprescribing/frontend/management/commands/infer_practice_boundaries.py

```diff
@@ -0,0 +1,130 @@
+"""
+Infer boundaries for practices from their locations using a Voronoi partition
+(i.e. assign each point in the country to its closest practice)
+
+These boundaries have no real meaning but they're easy enough to explain and
+they allow us to render plausible-looking maps for arbitrary collections of
+practices.
+
+The boundaries are clipped at the national border to stop them extending into
+the sea -- or Wales -- and generally looking ridiculous.
+"""
+from django.core.management.base import BaseCommand
+from django.db.models import Func
+from django.contrib.gis.db.models import Collect, Union
+from django.db import connection, transaction
+
+from frontend.models import Practice, PCT
+
+
+class Command(BaseCommand):
+
+    help = __doc__
+
+    def handle(self, *args, **options):
+        infer_practice_boundaries()
+
+
+def infer_practice_boundaries():
+    practices = (
+        Practice.objects
+        .filter(location__isnull=False, setting=4)
+        .exclude(
+            status_code__in=(
+                Practice.STATUS_RETIRED,
+                Practice.STATUS_DORMANT,
+                Practice.STATUS_CLOSED
+            )
+        )
+    )
+    partition = (
+        practices
+        .aggregate(
+            voronoi=Func(Collect('location'), function='ST_VoronoiPolygons')
+        )
+        ['voronoi']
+    )
+    national_boundary = get_national_boundary()
+    practice_regions = get_practice_code_to_region_map(partition, national_boundary)
+    with transaction.atomic():
+        for practice in practices:
+            practice.boundary = practice_regions[practice.code]
+            practice.save(update_fields=['boundary'])
+
+
+def get_practice_code_to_region_map(regions, clip_boundary):
+    with connection.cursor() as cursor:
+        return _get_practice_code_to_region_map(cursor, regions, clip_boundary)
+
+
+def _get_practice_code_to_region_map(cursor, regions, clip_boundary):
+    """
+    Return a dict mapping practice codes to the region in `regions` in which
+    they're located, with returned regions clipped to `clip_boundary`
+    """
+    cursor.execute(
+        'CREATE TEMPORARY TABLE regions (original GEOMETRY, clipped GEOMETRY)'
+    )
+    for region in regions:
+        clipped = region.intersection(clip_boundary)
+        # This is a workaround for the error "Relate Operation called with a
+        # LWGEOMCOLLECTION type" which happens when clipped boundaries end up
+        # including things which aren't polygons (i.e. points or lines) and we
```

Comment:
Out of interest, how or when would clipping a voronoi partition result in a line?